File: cakk-domain/mysql/src/main/java/com/cakk/domain/mysql/converter/LinkKindConverter.java
package com.cakk.domain.mysql.converter;

import java.util.stream.Stream;

import jakarta.persistence.AttributeConverter;

import com.cakk.common.enums.LinkKind;
import com.cakk.common.enums.ReturnCode;
import com.cakk.common.exception.CakkException;

public class LinkKindConverter implements AttributeConverter<LinkKind, String> {

	@Override
	public String convertToDatabaseColumn(final LinkKind linkKind) {
		if (linkKind == null) {
			return null;
		}

		return linkKind.getValue();
	}

	@Override
	public LinkKind convertToEntityAttribute(final String value) {
		if (value == null) {
			return null;
		}

		return Stream.of(LinkKind.values())
			.filter(linkKind -> linkKind.getValue().equals(value))
			.findFirst()
			.orElseThrow(() -> new CakkException(ReturnCode.INTERNAL_SERVER_ERROR));
	}
}

File: cakk-domain/mysql/src/main/java/com/cakk/domain/mysql/converter/LinkKindConverter.kt
package com.cakk.domain.mysql.converter

import com.cakk.common.enums.LinkKind
import com.cakk.common.enums.ReturnCode
import com.cakk.common.exception.CakkException
import jakarta.persistence.AttributeConverter
import java.util.stream.Stream

class LinkKindConverter : AttributeConverter<LinkKind?, String?> {

	override fun convertToDatabaseColumn(linkKind: LinkKind?): String? {
        return linkKind?.value
    }

    override fun convertToEntityAttribute(value: String?): LinkKind? {
        return if (value == null) {
            null
        } else Stream.of(*LinkKind.values())
                .filter { linkKind: LinkKind -> linkKind.value == value }
                .findFirst()
                .orElseThrow { CakkException(ReturnCode.INTERNAL_SERVER_ERROR) }
    }
}

